Ethkey - extended keys

NikVolf opened the pull request:
Deterministic derivation of private & public keys using secp256k1 elliptic curve

based on bitcoin hd wallet spec (BIP0032): https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki

gavofyork commented on ethkey/src/extended.rs:
```diff
@@ -0,0 +1,368 @@
+		let mut data = [0u8; 37];
+
+		let sec_private = SecretKey::from_slice(&SECP256K1, &*private_key)
+			.expect("Caller should provide valid private key");
```
not a valid expectation -> either prove or provide an error return path

NikVolf replied:
it's internal function, the upper code never passes random hashes in it

NikVolf replied:
added comment that it can panic just in case

gavofyork commented on ethkey/src/extended.rs:
```diff
@@ -0,0 +1,368 @@
+		// curve point (compressed public key) --  index
+		//             0.33                    --  33..37
+		data[0..33].copy_from_slice(&public_serialized);
+		BigEndian::write_u32(&mut data[33..37], index);
```
only u32? we might want to provide an extended version of this which can handle 256-bit `index`s

NikVolf replied:
why do you need so much?
using u32 allows to derive 2^31 soft and 2^31 hard keys

gavofyork replied:
yup, but that number is theoretically searchable. my understanding is that the maths allows for `2**256` soft keys, which then means that you can have cryptographically secure receive addresses. basically, you want to make a secret payment to address `X`: you generate address `Y = soft(X, H)` where `H` is some cryptographically secure invoice identifier (e.g. the `sha3` of the invoice PDF). if you don't have `H` then you cannot associate X with Y, but once you do have `H` then you can prove that the owner of `X` can access the funds at `Y`.

basically it lets you send funds to someone's public address without the payment itself being made public. but still with the ability for either party to make the payment public at a later date.

NikVolf replied:
Nice explanation, and It's easy to implement, actually

NikVolf replied:
#4438

gavofyork commented:
i presume those tests are the official test vectors

NikVolf commented:
@gavofyork 
there is no "official" test vectors for exactly this algorithm
i'll try to adapt bip0032 test vectors, but there can be errors while adapting as well :)

NikVolf commented:
@gavofyork 
done, extracted test vectors and added from bitcoin bip0032
it can be later published as our test vectors also once algorithm is somewhat proven good

gavofyork commented:
@NikVolf let me know if you can't see an easy way to extend index length to 256 bit.

gavofyork commented:
will merge as-is; 256-bit indexes can be added in a second PR.
